tools/test/test_depsolve.py

```diff
@@ -1409,29 +1522,135 @@ def test_depsolve(tmp_path, repo_servers, dnf_config, detect_fn, with_sbom, test
+    repo_configs = [gen_repo_config(server) for server in tc_repo_servers]
+
+    with TemporaryDirectory() as cache_dir:
+        res, exit_code = depsolve(transactions, cache_dir, dnf_config, repo_configs, with_sbom=with_sbom)
+
+        if test_case.get("error", False):
+            assert exit_code != 0
+            assert res["kind"] == test_case["error_kind"]
+            assert re.match(test_case["error_reason_re"], res["reason"], re.DOTALL)
+            return
+
+        assert exit_code == 0
+        assert {pkg["name"] for pkg in res["packages"]} == test_case["results"]["packages"]
+        assert res["repos"].keys() == test_case["results"]["reponames"]
+
+        for repo in res["repos"].values():
+            assert repo["gpgkeys"] == [TEST_KEY + repo["id"]]
+            assert repo["sslverify"] is False
+        if with_sbom:
+            assert "sbom" in res
+
+            spdx_2_3_1_schema_file = './test/data/spdx/spdx-schema-v2.3.1.json'
+            with open(spdx_2_3_1_schema_file, encoding="utf-8") as f:
+                spdx_schema = json.load(f)
+            validator = jsonschema.Draft4Validator
+            validator.check_schema(spdx_schema)
+            spdx_validator = validator(spdx_schema)
+            spdx_validator.validate(res["sbom"])
+
+            assert {pkg["name"] for pkg in res["sbom"]["packages"]} == test_case["results"]["packages"]
+        else:
+            assert "sbom" not in res
+
+        use_dnf5 = dnf_config.get("use_dnf5", False)
+        if use_dnf5:
+            assert res["solver"] == "dnf5"
+        else:
+            assert res["solver"] == "dnf"
+
+
+# pylint: disable=too-many-branches
+@pytest.mark.parametrize("test_case", depsolve_test_cases, ids=tcase_idfn)
+@pytest.mark.parametrize("dnf_config, detect_fn", [
+    ({}, assert_dnf),
+    ({"use_dnf5": False}, assert_dnf),
+    ({"use_dnf5": True}, assert_dnf5),
+], ids=["no-config", "dnf4", "dnf5"])
+def test_depsolve(repo_servers, dnf_config, detect_fn, test_case):
+    try:
+        detect_fn()
+    except RuntimeError as e:
+        pytest.skip(str(e))
+
+    # pylint: disable=fixme
+    # TODO: remove this once dnf5 implementation is fixed
+    dnf5_broken_test_cases = [
+        "basic_pkg_group_with_excludes",
+        "install_pkg_excluded_in_another_transaction",
+        "error_pkg_not_in_enabled_repos",
+        "basic_module",
+    ]
+
+    if dnf_config.get("use_dnf5", False) and test_case["id"] in dnf5_broken_test_cases:
+        pytest.skip("This test case is known to be broken with dnf5")
+
+    transactions = test_case["transactions"]
+
+    tc_repo_servers = get_test_case_repo_servers(test_case, repo_servers)
+
+    repo_configs = [gen_repo_config(server) for server in tc_repo_servers]
+    with TemporaryDirectory() as cache_dir:
+        res, exit_code = depsolve(transactions, cache_dir, dnf_config, repo_configs)
+
+        if test_case.get("error", False):
```

Comment:
See above, I wonder if there is an opportunity to extract a common helpers? It seems quite a bit of the validation overlaps with the test above